Introduce Hybrid vector clock

highcloud100 opened the pull request:
<!--  Thanks for sending a pull request! -->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Address #723

**Special notes for your reviewer**:

https://hackmd.io/kfaV_ZAgSX-REVyKRnoU1Q?view
https://hackmd.io/ogc6dOGkRW-At6O8SmiDrg?both#3-vector-clock%EC%9D%98-%ED%95%84%EC%9A%94%EC%84%B1

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note

```

**Additional documentation**:

<!--
This section can be blank if this pull request does not require a release note.

Please use the following format for linking documentation:
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

**Checklist**:
- [ ] Added relevant tests or not required
- [ ] Didn't break anything

highcloud100 committed:
Add gc_test

highcloud100 committed:
Add vector clock

1. Add synced vector map in document.
2. Add self vector clock in Change
3. Modify gc to work with min synced vector

highcloud100 committed:
Replace MaxTicket with MaxVectorClock in test

highcloud100 committed:
Fix some bugs

1. Old actor id does't convert to new in syncedVectorMap
2. Change the gc criteria to RemovedAt

highcloud100 committed:
Replace latestCreatedAtMapByActor with vector clock

1. replace text edit's  latestCreatedAtMapByActor with vector clock.
2. Fix gc timing in gc_test.go, because gc changed to use vectorclock
3. Because it is used in text edit, context includes current vector.

highcloud100 committed:
Lint

highcloud100 committed:
Fix MinSyncedVector and duplicate vectorClock updates

1. Fix MinSyncedVector and add relevant test.
2. Fix duplicate vectorClock updates due to pre `context` update commit.
3. Modify gc_test.go to fit hybrid vector clock, except detach test.

highcloud100 committed:
Lint and Structural change

highcloud100 committed:
Add synced vector map initialization using snapshots

1. server store synced vector map in snapshot_info.
2. The client also receives vector clock when pulling pack from the snapshot, and replace it's vector map with pulled vector map.

highcloud100 committed:
Fix bench test

hackerwins commented:
Related to https://github.com/yorkie-team/yorkie/issues/789

highcloud100 committed:
Optimize timing of calculating min seq vector

1. Calculate minimum vector only when garbage is present.
2. Copy depth when inserting vector clocks into Change.

hackerwins commented on pkg/document/time/vector_clock_test.go:
```diff
@@ -0,0 +1,62 @@
+		}, vec.MinSyncedVector())
+	})
+
+	t.Run("MaxVectorClock arbitary IDs test", func(t *testing.T) {
```
```suggestion
	t.Run("MaxVectorClock arbitrary IDs test", func(t *testing.T) {
```

hackerwins committed:
Update pkg/document/time/vector_clock_test.go

highcloud100 committed:
Add deleting vectorclock of detaced Actor from synced vector map

1. Now Change has detach flag. It used to delete vectorclock of detached Actor.

highcloud100 committed:
Fix snapshot just return the latestVectorClock not SVM

1. Now snapshot save latest vector clock not SVM.
2. Clients build their own SVM using the latest VectorClock in snapshots.
- So client's SVM is not latest state, it doesn't know peer's vector clock. It need more check
- This is due to the overhead of storing and transferring SVMs. #789

highcloud100 committed:
Delete Synced Vector Map

highcloud100 committed:
Add Calculate svm in server

This approach is silly.
- slow and behave like client.
- Server track all logs of clients, so we don't need calculate min synced vector every time.

highcloud100 committed:
Delete SVM from server, Add latest vector in change

highcloud100 committed:
Rename vector clocks and Fix gc_test

Terminology
- change's vector clock -> vector clock
- change's latest vector clock -> version vector
- each document's vector clock -> version vector
- min synced vector clock -> synced version vector
- snapshot's vector clock -> latest version vector

highcloud100 committed:
Fix bench code error

hackerwins commented:
This continues at #800.

hackerwins commented:
About Min and LCA:
https://codepair.yorkie.dev/highcloud100/66bd684f28737f961bccd807/share?token=pgt8es
